File: 07/src/tecs/vmtranslator/VMPattern.java
package tecs.vmtranslator;

class VMPattern {
	static final String PUSH = "\\s*push\\s+(\\S+)\\s+(\\d+)\\W*";
	static final String POP = "\\s*pop\\s+(\\S+)\\s+(\\d+)\\W*";
	static final String ARITHMETIC = "\\s*(add|sub)\\W*";

	// more to come
	static final String LABEL = "\\s*(add|sub)\\W*";
	static final String GOTO = "\\s*(add|sub)\\W*";
	static final String IF = "\\s*(add|sub)\\W*";
	static final String FUNCTION = "\\s*(add|sub)\\W*";
	static final String RETURN = "\\s*(add|sub)\\W*";
	static final String CALL = "\\s*(add|sub)\\W*";
}

File: 07/src/tecs/vmtranslator/VMPattern.kt
package tecs.vmtranslator

object VMPattern {
    // push local 0
    val PUSH = """\s*push\s+(\w+)\s+(\d+)\s*(//.*)?""".toRegex()

    // pop constant 10
    val POP = """\s*pop\s+(\S+)\s+(\d+)\s*(//.*)?""".toRegex()

    // add
    val ARITHMETIC = """\s*(add|sub|neg|eq|gt|lt|and|or|not)\s*(//.*)?""".toRegex()

    // label foo
    val LABEL = """\s*label\s+(.+)\s*(//.*)?""".toRegex()

    // goto foo
    val GOTO = """\s*goto\s+(.+)\s*(//.*)?""".toRegex()

    // if-goto foo
    val IF = """\s*if-goto\s+(.+)\s*(//.*)?""".toRegex()

    // function functionName nArgs
    val FUNCTION = """\s*function\s+(.+)\s+(\d+)\s*(//.*)?""".toRegex()

    // TODO
    val CALL = """\s*call\s+(.+)\s+(\d+)\s*(//.*)?""".toRegex()

    // return
    val RETURN = """\s*return\s*(//.*)?""".toRegex()
}
